src/hermes/jsi/jsi.h

```diff
@@ -0,0 +1,1326 @@
+  /// this.
+  Value callAsConstructor(Runtime& runtime, const Value* args, size_t count)
+      const;
+
+  /// Same as above `callAsConstructor`, except use an initializer_list to
+  /// supply the arguments.
+  Value callAsConstructor(Runtime& runtime, std::initializer_list<Value> args)
+      const;
+
+  /// Same as above `callAsConstructor`, but automatically converts/wraps
+  /// any argument with a jsi Value.
+  template <typename... Args>
+  Value callAsConstructor(Runtime& runtime, Args&&... args) const;
+
+  /// Returns whether this was created with Function::createFromHostFunction.
+  /// If true then you can use getHostFunction to get the underlying
+  /// HostFunctionType.
+  bool isHostFunction(Runtime& runtime) const {
+    return runtime.isHostFunction(*this);
+  }
+
+  /// Returns the underlying HostFunctionType iff isHostFunction returns true
+  /// and asserts otherwise. You can use this to use std::function<>::target
+  /// to get the object that was passed to create the HostFunctionType.
+  ///
+  /// Note: The reference returned is borrowed from the JS object underlying
+  ///       \c this, and thus only lasts as long as the object underlying
+  ///       \c this does.
+  HostFunctionType& getHostFunction(Runtime& runtime) const {
+    assert(isHostFunction(runtime));
+    return runtime.getHostFunction(*this);
+  }
+
+ private:
+  friend class Object;
+  friend class Value;
+
+  Function(Runtime::PointerValue* value) : Object(value) {}
+};
+
+/// Represents any JS Value (undefined, null, boolean, number, symbol,
+/// string, or object).  Movable, or explicitly copyable (has no copy
+/// ctor).
+class JSI_EXPORT Value {
+ public:
+  /// Default ctor creates an \c undefined JS value.
+  Value() : Value(UndefinedKind) {}
+
+  /// Creates a \c null JS value.
+  /* implicit */ Value(std::nullptr_t) : kind_(NullKind) {}
+
+  /// Creates a boolean JS value.
+  /* implicit */ Value(bool b) : Value(BooleanKind) {
+    data_.boolean = b;
+  }
+
+  /// Creates a number JS value.
+  /* implicit */ Value(double d) : Value(NumberKind) {
+    data_.number = d;
+  }
+
+  /// Creates a number JS value.
+  /* implicit */ Value(int i) : Value(NumberKind) {
+    data_.number = i;
+  }
+
+  /// Moves a Symbol, String, or Object rvalue into a new JS value.
+#if 0
+  template <typename T>
+#else // XXX Realm modification
+  template <typename T, typename = std::enable_if_t<
+            !std::is_reference_v<T> && 
+            (std::is_base_of<Symbol, T>::value ||
+             std::is_base_of<String, T>::value ||
+             std::is_base_of<Object, T>::value)
+          >>
+#endif
+  /* implicit */ Value(T&& other) : Value(kindOf(other)) {
+    static_assert(
+        std::is_base_of<Symbol, T>::value ||
+            std::is_base_of<String, T>::value ||
+            std::is_base_of<Object, T>::value,
+        "Value cannot be implicitly move-constructed from this type");
```

Comment:
For now I've modified this to move the error out of the function to overload resolution. This results in much better error messages when I screw up. Before considering this "done" I'll want to either replace this file with a clean copy from upstream, or ideal set up the build so that we just use theirs without our own copy. But for now, it is nice having this modification in place.

Also, I plan to open a PR to do this upstream, so maybe we can have the best of both worlds.

Reply:
FYI: https://github.com/facebook/hermes/pull/526